### PR TITLE
Add DingTalkJobErrorHandler to support sending message

### LINKED ISSUE #1460: Add DingTalkJobErrorHandler to support sending message
To support for sending the messages when the job executes with errors, we decide to do some extensions for JobErrorHandler.
- create a new module - elasticjob-error-handler/elasticjob-error-handler-dingtalk
- load config from System parameters.
- load config from specified conf.

### PR DESCRIPTION
**Pull request by gzdzss**
Fixes #1460.

Changes proposed in this pull request:
- Add DingTalkJobErrorHandler to support sending message

### CODE REVIEW

File: elasticjob-error-handler/elasticjob-error-handler-dingtalk/pom.xml [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+            <version>${project.version}</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <scope>test</scope>
+        </dependency>
+    </dependencies>
+</project>
```
Comment (TeslaCN) [inline elasticjob-error-handler/elasticjob-error-handler-dingtalk/pom.xml]: Need a blank line.

File: elasticjob-error-handler/elasticjob-error-handler-dingtalk/src/test/java/org/apache/shardingsphere/elasticjob/error/handler/impl/fixture/DingtalkInternalController.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+@Slf4j
+public final class DingtalkInternalController implements RestfulController {
+    
+    private static final String ACCESS_TOKEN = "42eead064e81ce81fc6af2c107fbe10a4339a3d40a7db8abf5b34d8261527a3f";
+    
+    private static final String KEYWORD = "keyword";
+    
+    private static final String SECRET = "SEC0b0a6b13b6823b95737dd83491c23adee5d8a7a649899a12217e038eddc84ff4";
+    
+
```
Comment (TeslaCN) [inline elasticjob-error-handler/elasticjob-error-handler-dingtalk/src/test/java/org/apache/shardingsphere/elasticjob/error/handler/impl/fixture/DingtalkInternalController.java]: Remove redundant blank line.

Comment (gzdzss) [reply elasticjob-error-handler/elasticjob-error-handler-dingtalk/src/test/java/org/apache/shardingsphere/elasticjob/error/handler/impl/fixture/DingtalkInternalController.java]: got it

File: elasticjob-error-handler/elasticjob-error-handler-dingtalk/pom.xml [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+    <dependencies>
+        <dependency>
+            <groupId>org.apache.shardingsphere.elasticjob</groupId>
+            <artifactId>elasticjob-error-handler-general</artifactId>
+            <version>${project.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>org.projectlombok</groupId>
+            <artifactId>lombok</artifactId>
```
Comment (TeslaCN) [inline elasticjob-error-handler/elasticjob-error-handler-dingtalk/pom.xml:45]: "Provided" scope is recommended.

File: elasticjob-error-handler/elasticjob-error-handler-dingtalk/pom.xml [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+            <artifactId>logback-classic</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.slf4j</groupId>
+            <artifactId>log4j-over-slf4j</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>org.slf4j</groupId>
+            <artifactId>jcl-over-slf4j</artifactId>
+        </dependency>
```
Comment (TeslaCN) [inline elasticjob-error-handler/elasticjob-error-handler-dingtalk/pom.xml]: Are they optional?

**Review by TeslaCN — CHANGES_REQUESTED** [review]

File: elasticjob-error-handler/elasticjob-error-handler-dingtalk/src/main/java/org/apache/shardingsphere/elasticjob/error/handler/impl/DingtalkJobErrorHandler.java [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+                    log.error("An exception has occurred in Job '{}', But failed to send alert by Dingtalk because of: {}", jobName, resp.get("errmsg").getAsString(), cause);
+                } else {
+                    log.error("An exception has occurred in Job '{}', Notification to Dingtalk was successful.", jobName, cause);
+                }
+            } else {
+                log.error("An exception has occurred in Job '{}', But failed to send alert by Dingtalk because of: Unexpected response status: {}", jobName, status, cause);
+            }
+        } finally {
+            response.close();
+        }
```
Comment (TeslaCN) [inline elasticjob-error-handler/elasticjob-error-handler-dingtalk/src/main/java/org/apache/shardingsphere/elasticjob/error/handler/impl/DingtalkJobErrorHandler.java]: CloseableHttpResponse implement AutoCloseable. Replace finally with try-with-resources.

File: elasticjob-error-handler/elasticjob-error-handler-dingtalk/src/main/java/org/apache/shardingsphere/elasticjob/error/handler/impl/DingtalkJobErrorHandler.java [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        if (null == dingtalkConfiguration) {
+            dingtalkConfiguration = DingtalkEnvironment.getInstance().getDingtalkConfiguration();
+        }
+        HttpPost httpPost = new HttpPost(getUrl());
+        RequestConfig requestConfig = RequestConfig.custom()
+                .setConnectTimeout(dingtalkConfiguration.getConnectTimeout())
+                .setSocketTimeout(dingtalkConfiguration.getReadTimeout()).build();
+        httpPost.setConfig(requestConfig);
+        String paramJson = getParamJson(getMsg(jobName, cause));
+        StringEntity entity = new StringEntity(paramJson, StandardCharsets.UTF_8.name());
```
Comment (TeslaCN) [inline elasticjob-error-handler/elasticjob-error-handler-dingtalk/src/main/java/org/apache/shardingsphere/elasticjob/error/handler/impl/DingtalkJobErrorHandler.java]: Just use "StandardCharsets.UTF_8". The "name()" is unnecessary.

File: elasticjob-error-handler/elasticjob-error-handler-dingtalk/src/main/java/org/apache/shardingsphere/elasticjob/error/handler/impl/DingtalkJobErrorHandler.java [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+/**
+ * Job error handler for dingtalk error message.
+ */
+@Slf4j
+public final class DingtalkJobErrorHandler implements JobErrorHandler {
+    
+    @Setter
+    private DingtalkConfiguration dingtalkConfiguration;
+    
+    private CloseableHttpClient httpclient = HttpClients.createDefault();
```
Comment (TeslaCN) [inline elasticjob-error-handler/elasticjob-error-handler-dingtalk/src/main/java/org/apache/shardingsphere/elasticjob/error/handler/impl/DingtalkJobErrorHandler.java]: This field may be "final".

**Review by TeslaCN — CHANGES_REQUESTED** [review]

File: elasticjob-error-handler/elasticjob-error-handler-dingtalk/src/main/java/org/apache/shardingsphere/elasticjob/error/handler/impl/DingtalkJobErrorHandler.java [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+    private String getSignUrl() throws NoSuchAlgorithmException, InvalidKeyException, UnsupportedEncodingException {
+        Long timestamp = System.currentTimeMillis();
+        return String.format("%s&timestamp=%s&sign=%s", dingtalkConfiguration.getWebhook(), timestamp, sign(timestamp));
+    }
+    
+    private String sign(final Long timestamp) throws NoSuchAlgorithmException, UnsupportedEncodingException, InvalidKeyException {
+        String stringToSign = timestamp + "\n" + dingtalkConfiguration.getSecret();
+        Mac mac = Mac.getInstance("HmacSHA256");
+        mac.init(new SecretKeySpec(dingtalkConfiguration.getSecret().getBytes(StandardCharsets.UTF_8.name()), "HmacSHA256"));
+        byte[] signData = mac.doFinal(stringToSign.getBytes(StandardCharsets.UTF_8.name()));
```
Comment (TeslaCN) [inline elasticjob-error-handler/elasticjob-error-handler-dingtalk/src/main/java/org/apache/shardingsphere/elasticjob/error/handler/impl/DingtalkJobErrorHandler.java]: Just use "StandardCharsets.UTF_8". The "name()" is unnecessary for "getBytes".

File: elasticjob-error-handler/elasticjob-error-handler-dingtalk/src/main/java/org/apache/shardingsphere/elasticjob/error/handler/impl/DingtalkJobErrorHandler.java [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+        StringWriter sw = new StringWriter();
+        cause.printStackTrace(new PrintWriter(sw, true));
+        String msg = String.format("Job '%s' exception occur in job processing, caused by %s", jobName, sw.toString());
+        if (!Strings.isNullOrEmpty(dingtalkConfiguration.getKeyword())) {
+            msg = dingtalkConfiguration.getKeyword().concat(msg);
+        }
+        return msg;
+    }
+    
+    private String getUrl() throws NoSuchAlgorithmException, InvalidKeyException, UnsupportedEncodingException, MalformedURLException {
```
Comment (TeslaCN) [inline elasticjob-error-handler/elasticjob-error-handler-dingtalk/src/main/java/org/apache/shardingsphere/elasticjob/error/handler/impl/DingtalkJobErrorHandler.java]: MalformedURLException can be removed.

File: elasticjob-error-handler/elasticjob-error-handler-dingtalk/pom.xml [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+  ~ limitations under the License.
+  -->
+
+<project xmlns="http://maven.apache.org/POM/4.0.0"
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <artifactId>elasticjob-error-handler</artifactId>
+        <groupId>org.apache.shardingsphere.elasticjob</groupId>
```
Comment (Technoboy-) [inline elasticjob-error-handler/elasticjob-error-handler-dingtalk/pom.xml:24]: <groupId> first, and then artifactId

File: elasticjob-error-handler/elasticjob-error-handler-dingtalk/pom.xml [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+        <dependency>
+            <groupId>org.mockito</groupId>
+            <artifactId>mockito-core</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.shardingsphere.elasticjob</groupId>
+            <artifactId>elasticjob-restful</artifactId>
+            <version>${project.version}</version>
+            <scope>test</scope>
```
Comment (Technoboy-) [inline elasticjob-error-handler/elasticjob-error-handler-dingtalk/pom.xml:77]: If relys on ejob dependency, move it head.  Before lombok

File: elasticjob-error-handler/elasticjob-error-handler-dingtalk/pom.xml [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+    <dependencies>
+        <dependency>
+            <groupId>org.apache.shardingsphere.elasticjob</groupId>
+            <artifactId>elasticjob-error-handler-general</artifactId>
+            <version>${project.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>org.projectlombok</groupId>
+            <artifactId>lombok</artifactId>
+            <scope>provided</scope>
```
Comment (Technoboy-) [inline elasticjob-error-handler/elasticjob-error-handler-dingtalk/pom.xml]: No provided need.

File: elasticjob-error-handler/elasticjob-error-handler-dingtalk/pom.xml [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+        </dependency>
+        <dependency>
+            <groupId>org.apache.httpcomponents</groupId>
+            <artifactId>httpcore</artifactId>
+            <version>4.4.13</version>
+        </dependency>
+        <dependency>
+            <groupId>org.mockito</groupId>
+            <artifactId>mockito-core</artifactId>
+            <scope>test</scope>
```
Comment (Technoboy-) [inline elasticjob-error-handler/elasticjob-error-handler-dingtalk/pom.xml:72]: No need to add test

File: elasticjob-error-handler/elasticjob-error-handler-dingtalk/pom.xml [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+        <dependency>
+            <groupId>org.apache.shardingsphere.elasticjob</groupId>
+            <artifactId>elasticjob-restful</artifactId>
+            <version>${project.version}</version>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <scope>test</scope>
```
Comment (Technoboy-) [inline elasticjob-error-handler/elasticjob-error-handler-dingtalk/pom.xml:82]: No need to add.

File: elasticjob-error-handler/elasticjob-error-handler-dingtalk/src/main/java/org/apache/shardingsphere/elasticjob/error/handler/config/DingtalkConfiguration.java [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+ */
+
+package org.apache.shardingsphere.elasticjob.error.handler.config;
+
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import lombok.Setter;
+
+@Getter
+@Setter
```
Comment (Technoboy-) [inline elasticjob-error-handler/elasticjob-error-handler-dingtalk/src/main/java/org/apache/shardingsphere/elasticjob/error/handler/config/DingtalkConfiguration.java]: No need setter

**Review by Technoboy- — APPROVED** [review]

**Pull request merged**